images/virtualization-artifact/pkg/controller/vmip/vmip_webhook.go

```diff
@@ -0,0 +1,156 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vmip
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmip/internal/util"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vmipcondition"
+)
+
+func NewValidator(log logr.Logger, client client.Client) *Validator {
+	return &Validator{log: log.WithName(controllerName).WithValues("webhook", "validation"), client: client}
+}
+
+type Validator struct {
+	log    logr.Logger
+	client client.Client
+}
+
+func (v *Validator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	vmip, ok := obj.(*v1alpha2.VirtualMachineIPAddress)
+	if !ok {
+		return nil, fmt.Errorf("expected a new VirtualMachineIPAddress but got a %T", obj)
+	}
+
+	v.log.Info("Validate VirtualMachineIP creating", "name", vmip.Name, "type", vmip.Spec.Type, "address", vmip.Spec.StaticIP)
+
+	err := v.validateSpecFields(vmip.Spec)
+	if err != nil {
+		return nil, fmt.Errorf("error validating VirtualMachineIP creation: %w", err)
+	}
+
+	if vmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeStatic {
+		ip := vmip.Spec.StaticIP
+
+		if ip != "" {
+			allocatedIPs, err := util.GetAllocatedIPs(ctx, v.client, vmip.Spec.Type)
+			if err != nil {
+				return nil, fmt.Errorf("error getting allocated ips: %w", err)
+			}
+
+			allocatedLease, ok := allocatedIPs[ip]
+			if ok && allocatedLease.Spec.VirtualMachineIPAddressRef != nil &&
+				(allocatedLease.Spec.VirtualMachineIPAddressRef.Namespace != vmip.Namespace ||
+					allocatedLease.Spec.VirtualMachineIPAddressRef.Name != vmip.Name) {
+				return nil, fmt.Errorf("VirtualMachineIPAddress cannot be created: the IP address %s has already been allocated by VirtualMachineIPAddress/%s in ns/%s", ip, allocatedLease.Spec.VirtualMachineIPAddressRef.Name, allocatedLease.Spec.VirtualMachineIPAddressRef.Namespace)
+			}
+		}
+	}
+
+	if vmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeAuto && vmip.Spec.StaticIP != "" {
+		v.log.Error(nil, "Invalid combination: StaticIP is specified with Type 'Auto'. The StaticIP will be ignored.", "VirtualMachineIPName", vmip.Name, "StaticIP", vmip.Spec.StaticIP)
```

Comment:
This is not internal error. It's better to return warning here

Reply:
You're probably right. Fixed it.